### PR TITLE
[MAT-255] Add materialisation to OGTerminal

### PR DESCRIPTION
**Pull request by gmarkall**
This adds materialisation to OGTerminal types, for scalars, dense
and diagonal matrices.

toReal16Array and toComplex16Array methods are added to
OGTerminal, which are presently implemented for diagonal matrices
and are used to materialise them efficiently.

Functions for creating Java arrays of arrays are moved into
dispatch.cc, since they are reused by the dispatcher that
materialises OGTerminals. Methods for constructing Java arrays of
doubles are also added.

Librdag code coverage: 100%

Buildbot pass: http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/335

### CODE REVIEW

File: src/jshim/dispatch.cc [hunk 1]
```diff
@@ -201,4 +209,248 @@ int DispatchToComplex16ArrayOfArrays::getCols()
...
+}
+
+
+/*
+ * Make a new double object because there's no autoboxing in JNI.
+ */
+
+jobject NewDouble(JNIEnv* env, jdouble v)
+{
+  jclass cls = env->FindClass("java/lang/Double");
```
Comment (stuartarchibald) [inline src/jshim/dispatch.cc]: isn't this somewhere in the JVMManager cache? if not it should be.

File: src/jshim/dispatch.cc [hunk 2]
```diff
@@ -201,4 +209,248 @@ int DispatchToComplex16ArrayOfArrays::getCols()
...
+  jmethodID constructor = env->GetMethodID(cls, "<init>", "(D)V");
```
Comment (stuartarchibald) [inline src/jshim/dispatch.cc]: Same goes for this.

File: src/jshim/dispatch.cc [hunk 3]
```diff
@@ -201,4 +209,248 @@ int DispatchToComplex16ArrayOfArrays::getCols()
...
+DispatchToOGTerminal::visit(librdag::OGExpr SUPPRESS_UNUSED const *thing)
+{
+  throw convert_error("Dispatch of expression to terminal not allowed");
+}
+
+void
+DispatchToOGTerminal::visit(librdag::OGScalar<real16> const *thing)
+{
+  jclass cls = _env->FindClass("com/opengamma/longdog/datacontainers/scalar/OGRealScalar");
+  jmethodID constructor = _env->GetMethodID(cls, "<init>", "(Ljava/lang/Number;)V");
```
Comment (stuartarchibald) [inline src/jshim/dispatch.cc]: and this

File: src/jshim/dispatch.cc (hunk 1 above)
Comment (gmarkall) [inline src/jshim/dispatch.cc]: OK - if it's not there I'll put it there. If it is, I'll use it.

**Comment by gmarkall** [comment]
I believe that I've now addressed caching of the classes and their constructors, and checking for exceptions before calling `entrypt()`. 

New buildbot pass: http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/341

Coverage of librdag still at 100% and Java coverage unchanged.

File: src/jshim/jshim.cc [hunk 4]
```diff
@@ -18,79 +18,22 @@
...
- * Converts a real16 ** to a java double[][]
- * @param env, the JNI environment pointer
- * @param inputData, the real16 array of arrays to convert
- * @param rows the number of rows in the array
- * @param cols the number of columns in the array
- * @return a jobjectArray which is a double[][] equivalent of {@code inputData}
+ * Check for exception
  */
-jobjectArray convertCreal16ArrOfArr2JDoubleArrOfArr(JNIEnv * env, real16 ** inputData, int rows, int cols)
+void checkEx(JNIEnv* env)
```
Comment (stuartarchibald) [inline src/jshim/jshim.cc:24]: why did git decide to do this!?!?!

**Comment by stuartarchibald** [comment]
Added MAT-261 to implement missing materialisations of sparse matrix types.

**Comment by stuartarchibald** [comment]
Happy to merge as WIP else. Good job.

**Pull request merged**